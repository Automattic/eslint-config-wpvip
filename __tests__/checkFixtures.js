import { ESLint } from 'eslint';

async function getLintMessages( type, fixture ) {
	const eslint = new ESLint( {
		overrideConfigFile: `__fixtures__/${ type }/eslintrc.js`,
		useEslintrc: false,
	} );

	const files = `__fixtures__/${ type }/${ fixture }`;
	const [ { messages } ] = await eslint.lintFiles( files );

	return messages;
}

describe( 'linting', () => {
	it.each( [
		[ 'base', 'allowed.js' ],
		[ 'cli', 'allowed.js' ],
		[ 'typescript', 'allowed.ts' ],
	] )( '%s %s fixture produces no errors', async ( type, fixture ) => {
		// Allowed fixtures should produce no lint messages.
		expect( await getLintMessages( type, fixture ) ).toStrictEqual( [] );
	} );

	it.each( [
		[ 'base', 'disallowed.js' ],
		[ 'cli', 'disallowed.js' ],
		[ 'typescript', 'disallowed.ts' ],
	] )( '%s %s fixture matches snapshot', async ( type, fixture ) => {
		// Disallowed fixtures produce messages that should match a snapshot.
		expect( await getLintMessages( type, fixture ) ).toMatchSnapshot();
	} );
} );
