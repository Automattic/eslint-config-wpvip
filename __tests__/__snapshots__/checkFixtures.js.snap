// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linting base disallowed.js fixture matches snapshot 1`] = `
Array [
  Object {
    "column": 12,
    "endColumn": 38,
    "endLine": 1,
    "line": 1,
    "message": "Found require(\\"child_process\\")",
    "nodeType": "CallExpression",
    "ruleId": "security/detect-child-process",
    "severity": 1,
  },
  Object {
    "column": 1,
    "endColumn": 8,
    "endLine": 4,
    "line": 4,
    "message": "Found child_process.exec() with non Literal first argument",
    "nodeType": "MemberExpression",
    "ruleId": "security/detect-child-process",
    "severity": 1,
  },
  Object {
    "column": 1,
    "endColumn": 26,
    "endLine": 7,
    "line": 7,
    "message": "Found non-literal argument in require",
    "nodeType": "CallExpression",
    "ruleId": "security/detect-non-literal-require",
    "severity": 1,
  },
  Object {
    "column": 1,
    "endColumn": 25,
    "endLine": 10,
    "line": 10,
    "message": "Found non-literal argument to RegExp Constructor",
    "nodeType": "NewExpression",
    "ruleId": "security/detect-non-literal-regexp",
    "severity": 1,
  },
  Object {
    "column": 10,
    "endColumn": 11,
    "endLine": 12,
    "line": 12,
    "message": "Identifier name 'x' is too short (< 2).",
    "messageId": "tooShort",
    "nodeType": "Identifier",
    "ruleId": "id-length",
    "severity": 1,
  },
  Object {
    "column": 10,
    "endColumn": 11,
    "endLine": 12,
    "line": 12,
    "message": "'x' is defined but never used.",
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "no-unused-vars",
    "severity": 2,
  },
  Object {
    "column": 13,
    "endColumn": 14,
    "endLine": 12,
    "line": 12,
    "message": "Identifier name 'y' is too short (< 2).",
    "messageId": "tooShort",
    "nodeType": "Identifier",
    "ruleId": "id-length",
    "severity": 1,
  },
  Object {
    "column": 13,
    "endColumn": 14,
    "endLine": 12,
    "line": 12,
    "message": "'y' is defined but never used.",
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "no-unused-vars",
    "severity": 2,
  },
  Object {
    "column": 9,
    "endColumn": 22,
    "endLine": 13,
    "line": 13,
    "message": "'xIsACoolParam' is not defined.",
    "messageId": "undef",
    "nodeType": "Identifier",
    "ruleId": "no-undef",
    "severity": 2,
  },
  Object {
    "column": 2,
    "endColumn": 3,
    "endLine": 14,
    "fix": Object {
      "range": Array [
        297,
        304,
      ],
      "text": "}

let",
    },
    "line": 14,
    "message": "Unnecessary semicolon.",
    "messageId": "unexpected",
    "nodeType": "EmptyStatement",
    "ruleId": "no-extra-semi",
    "severity": 2,
  },
  Object {
    "column": 5,
    "endColumn": 11,
    "endLine": 16,
    "fix": Object {
      "range": Array [
        301,
        427,
      ],
      "text": "const things = {
	a: 13,
	cd: 99.99,
	e9: 6e99,
	f7: (  ) => ['x'],
	12: '32',
	ff: xx => xx,
	gg: ( {fds, }) =>
        fds,
};",
    },
    "line": 16,
    "message": "'things' is never reassigned. Use 'const' instead.",
    "messageId": "useConst",
    "nodeType": "Identifier",
    "ruleId": "prefer-const",
    "severity": 2,
  },
  Object {
    "column": 2,
    "endColumn": 3,
    "endLine": 17,
    "line": 17,
    "message": "Identifier name 'a' is too short (< 2).",
    "messageId": "tooShort",
    "nodeType": "Identifier",
    "ruleId": "id-length",
    "severity": 1,
  },
  Object {
    "column": 7,
    "endColumn": 9,
    "endLine": 20,
    "fix": Object {
      "range": Array [
        353,
        355,
      ],
      "text": "",
    },
    "line": 20,
    "message": "There should be no space after this paren.",
    "messageId": "rejectedOpeningSpace",
    "nodeType": "Program",
    "ruleId": "space-in-parens",
    "severity": 2,
  },
  Object {
    "column": 7,
    "endColumn": 9,
    "endLine": 20,
    "fix": Object {
      "range": Array [
        353,
        355,
      ],
      "text": "",
    },
    "line": 20,
    "message": "There should be no space before this paren.",
    "messageId": "rejectedClosingSpace",
    "nodeType": "Program",
    "ruleId": "space-in-parens",
    "severity": 2,
  },
  Object {
    "column": 7,
    "endColumn": 9,
    "endLine": 20,
    "fix": Object {
      "range": Array [
        353,
        355,
      ],
      "text": " ",
    },
    "line": 20,
    "message": "Multiple spaces found before ')'.",
    "messageId": "multipleSpaces",
    "nodeType": "Punctuator",
    "ruleId": "no-multi-spaces",
    "severity": 2,
  },
  Object {
    "column": 14,
    "endColumn": 15,
    "endLine": 20,
    "fix": Object {
      "range": Array [
        361,
        361,
      ],
      "text": " ",
    },
    "line": 20,
    "message": "A space is required after '['.",
    "messageId": "missingSpaceAfter",
    "nodeType": "ArrayExpression",
    "ruleId": "array-bracket-spacing",
    "severity": 2,
  },
  Object {
    "column": 18,
    "endColumn": 19,
    "endLine": 20,
    "fix": Object {
      "range": Array [
        364,
        364,
      ],
      "text": " ",
    },
    "line": 20,
    "message": "A space is required before ']'.",
    "messageId": "missingSpaceBefore",
    "nodeType": "ArrayExpression",
    "ruleId": "array-bracket-spacing",
    "severity": 2,
  },
  Object {
    "column": 8,
    "endColumn": 9,
    "endLine": 23,
    "fix": Object {
      "range": Array [
        401,
        401,
      ],
      "text": " ",
    },
    "line": 23,
    "message": "A space is required after '{'.",
    "messageId": "requireSpaceAfter",
    "nodeType": "ObjectPattern",
    "ruleId": "object-curly-spacing",
    "severity": 2,
  },
  Object {
    "column": 12,
    "endColumn": 13,
    "endLine": 23,
    "fix": Object {
      "range": Array [
        401,
        407,
      ],
      "text": "fds }",
    },
    "line": 23,
    "message": "Unexpected trailing comma.",
    "messageId": "unexpected",
    "nodeType": "Property",
    "ruleId": "comma-dangle",
    "severity": 2,
  },
  Object {
    "column": 15,
    "endColumn": 16,
    "endLine": 23,
    "fix": Object {
      "range": Array [
        407,
        407,
      ],
      "text": " ",
    },
    "line": 23,
    "message": "There must be a space before this paren.",
    "messageId": "missingClosingSpace",
    "nodeType": "Program",
    "ruleId": "space-in-parens",
    "severity": 2,
  },
  Object {
    "column": 1,
    "endColumn": 9,
    "endLine": 24,
    "fix": Object {
      "range": Array [
        412,
        420,
      ],
      "text": "		",
    },
    "line": 24,
    "message": "Expected indentation of 2 tabs but found 8 spaces.",
    "messageId": "wrongIndentation",
    "nodeType": "Identifier",
    "ruleId": "indent",
    "severity": 2,
  },
  Object {
    "column": 3,
    "endColumn": 33,
    "endLine": 29,
    "line": 29,
    "message": "Unexpected \`await\` inside a loop.",
    "messageId": "unexpectedAwait",
    "nodeType": "AwaitExpression",
    "ruleId": "no-await-in-loop",
    "severity": 2,
  },
  Object {
    "column": 2,
    "endColumn": 6,
    "endLine": 36,
    "line": 34,
    "message": "No async function in forEachs",
    "nodeType": "ExpressionStatement",
    "ruleId": "no-async-foreach/no-async-foreach",
    "severity": 2,
  },
  Object {
    "column": 41,
    "endColumn": 46,
    "endLine": 34,
    "fix": Object {
      "range": Array [
        653,
        662,
      ],
      "text": "thing",
    },
    "line": 34,
    "message": "Unexpected parentheses around single function argument.",
    "messageId": "unexpectedParens",
    "nodeType": "ArrowFunctionExpression",
    "ruleId": "arrow-parens",
    "severity": 1,
  },
  Object {
    "column": 8,
    "endColumn": 11,
    "endLine": 40,
    "line": 40,
    "message": "'lol' is already declared in the upper scope on line 3 column 7.",
    "messageId": "noShadow",
    "nodeType": "Identifier",
    "ruleId": "no-shadow",
    "severity": 2,
  },
  Object {
    "column": 1,
    "endColumn": 12,
    "endLine": 44,
    "line": 44,
    "message": "'nonExistent' is not defined.",
    "messageId": "undef",
    "nodeType": "Identifier",
    "ruleId": "no-undef",
    "severity": 2,
  },
  Object {
    "column": 121,
    "fix": Object {
      "range": Array [
        906,
        906,
      ],
      "text": "
",
    },
    "line": 46,
    "message": "Newline required at end of file but not found.",
    "messageId": "missing",
    "nodeType": "Program",
    "ruleId": "eol-last",
    "severity": 2,
  },
]
`;

exports[`linting cli disallowed.js fixture matches snapshot 1`] = `Array []`;

exports[`linting prettier disallowed.js fixture matches snapshot 1`] = `
Array [
  Object {
    "column": 13,
    "endColumn": 18,
    "endLine": 5,
    "fix": Object {
      "range": Array [
        61,
        66,
      ],
      "text": "'x'",
    },
    "line": 5,
    "message": "Replace \`·'x'·\` with \`'x'\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 6,
    "endColumn": 8,
    "endLine": 7,
    "fix": Object {
      "range": Array [
        85,
        87,
      ],
      "text": "(xx)",
    },
    "line": 7,
    "message": "Replace \`xx\` with \`(xx)\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 7,
    "endColumn": 24,
    "endLine": 8,
    "fix": Object {
      "range": Array [
        101,
        118,
      ],
      "text": "{ fds }) => fds,",
    },
    "line": 8,
    "message": "Replace \`·{·fds·}·)·=>·fds\` with \`{·fds·})·=>·fds,\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 6,
    "endColumn": 44,
    "endLine": 13,
    "fix": Object {
      "range": Array [
        146,
        184,
      ],
      "text": "const thing of Object.keys(things)",
    },
    "line": 13,
    "message": "Replace \`·const·thing·of·Object.keys(·things·)·\` with \`const·thing·of·Object.keys(things)\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 1,
    "endColumn": 20,
    "endLine": 14,
    "fix": Object {
      "range": Array [
        188,
        207,
      ],
      "text": "	done.push(thing",
    },
    "line": 14,
    "message": "Replace \`··done.push(·thing·\` with \`↹done.push(thing\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
  Object {
    "column": 1,
    "endColumn": 1,
    "endLine": 21,
    "fix": Object {
      "range": Array [
        242,
        243,
      ],
      "text": "",
    },
    "line": 20,
    "message": "Delete \`⏎\`",
    "messageId": "delete",
    "nodeType": null,
    "ruleId": "prettier/prettier",
    "severity": 2,
  },
]
`;

exports[`linting typescript disallowed.ts fixture matches snapshot 1`] = `
Array [
  Object {
    "column": 48,
    "endColumn": 61,
    "endLine": 1,
    "line": 1,
    "message": "'three' is defined but never used. Allowed unused args must match /^_/u.",
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/no-unused-vars",
    "severity": 2,
  },
  Object {
    "column": 55,
    "endColumn": 61,
    "endLine": 1,
    "fix": Object {
      "range": Array [
        54,
        60,
      ],
      "text": "string",
    },
    "line": 1,
    "message": "Don't use \`String\` as a type. Use string instead",
    "messageId": "bannedTypeMessage",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/ban-types",
    "severity": 2,
  },
  Object {
    "column": 1,
    "endColumn": 14,
    "endLine": 5,
    "line": 5,
    "message": "Do not use \\"@ts-ignore\\" because it alters compilation errors.",
    "messageId": "tsDirectiveComment",
    "nodeType": "Line",
    "ruleId": "@typescript-eslint/ban-ts-comment",
    "severity": 2,
  },
  Object {
    "column": 45,
    "endColumn": 48,
    "endLine": 16,
    "line": 16,
    "message": "Unexpected any. Specify a different type.",
    "messageId": "unexpectedAny",
    "nodeType": "TSAnyKeyword",
    "ruleId": "@typescript-eslint/no-explicit-any",
    "severity": 1,
    "suggestions": Array [
      Object {
        "desc": "Use \`unknown\` instead, this will force you to explicitly, and safely assert the type is correct.",
        "fix": Object {
          "range": Array [
            289,
            292,
          ],
          "text": "unknown",
        },
        "messageId": "suggestUnknown",
      },
      Object {
        "desc": "Use \`never\` instead, this is useful when instantiating generic type parameters that you don't need to know the type of.",
        "fix": Object {
          "range": Array [
            289,
            292,
          ],
          "text": "never",
        },
        "messageId": "suggestNever",
      },
    ],
  },
  Object {
    "column": 8,
    "endColumn": 11,
    "endLine": 21,
    "line": 21,
    "message": "'add' is already declared in the upper scope.",
    "messageId": "noShadow",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/no-shadow",
    "severity": 2,
  },
]
`;

exports[`linting typescript-strict disallowed.ts fixture matches snapshot 1`] = `
Array [
  Object {
    "column": 48,
    "endColumn": 61,
    "endLine": 1,
    "line": 1,
    "message": "'three' is defined but never used. Allowed unused args must match /^_/u.",
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/no-unused-vars",
    "severity": 2,
  },
  Object {
    "column": 55,
    "endColumn": 61,
    "endLine": 1,
    "fix": Object {
      "range": Array [
        54,
        60,
      ],
      "text": "string",
    },
    "line": 1,
    "message": "Don't use \`String\` as a type. Use string instead",
    "messageId": "bannedTypeMessage",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/ban-types",
    "severity": 2,
  },
  Object {
    "column": 45,
    "endColumn": 48,
    "endLine": 5,
    "line": 5,
    "message": "Unexpected any. Specify a different type.",
    "messageId": "unexpectedAny",
    "nodeType": "TSAnyKeyword",
    "ruleId": "@typescript-eslint/no-explicit-any",
    "severity": 2,
    "suggestions": Array [
      Object {
        "desc": "Use \`unknown\` instead, this will force you to explicitly, and safely assert the type is correct.",
        "fix": Object {
          "range": Array [
            139,
            142,
          ],
          "text": "unknown",
        },
        "messageId": "suggestUnknown",
      },
      Object {
        "desc": "Use \`never\` instead, this is useful when instantiating generic type parameters that you don't need to know the type of.",
        "fix": Object {
          "range": Array [
            139,
            142,
          ],
          "text": "never",
        },
        "messageId": "suggestNever",
      },
    ],
  },
  Object {
    "column": 1,
    "endColumn": 14,
    "endLine": 9,
    "line": 9,
    "message": "Do not use \\"@ts-ignore\\" because it alters compilation errors.",
    "messageId": "tsDirectiveComment",
    "nodeType": "Line",
    "ruleId": "@typescript-eslint/ban-ts-comment",
    "severity": 2,
  },
  Object {
    "column": 21,
    "endColumn": 24,
    "endLine": 17,
    "line": 17,
    "message": "Unsafe return of an \`any\` typed value.",
    "messageId": "unsafeReturn",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/no-unsafe-return",
    "severity": 2,
  },
  Object {
    "column": 7,
    "endColumn": 20,
    "endLine": 20,
    "line": 20,
    "message": "'outerScopeVar' is assigned a value but never used.",
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/no-unused-vars",
    "severity": 2,
  },
  Object {
    "column": 10,
    "endColumn": 16,
    "endLine": 21,
    "line": 21,
    "message": "'shadow' is defined but never used.",
    "messageId": "unusedVar",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/no-unused-vars",
    "severity": 2,
  },
  Object {
    "column": 8,
    "endColumn": 21,
    "endLine": 22,
    "line": 22,
    "message": "'outerScopeVar' is already declared in the upper scope.",
    "messageId": "noShadow",
    "nodeType": "Identifier",
    "ruleId": "@typescript-eslint/no-shadow",
    "severity": 2,
  },
]
`;
